src/gt4py/next/otf/compiled_program.py

```diff
@@ -0,0 +1,275 @@
+        except TypeError:  # 'Future' object is not callable
+            # ... otherwise we resolve the future and call again
+            program = self._resolve_future(key)
+            program(*args, **kwargs, offset_provider=offset_provider)
+        except KeyError as e:
+            if enable_jit:
+                assert self.static_params is not None
+                static_args = {name: args[self._param_index(name)] for name in self.static_params}
+                self._compile_variant(static_args=static_args, offset_provider=offset_provider)
+                return self(
+                    *args, offset_provider=offset_provider, enable_jit=False, **kwargs
+                )  # passing `enable_jit=False` because a cache miss should be a hard-error in this call`
+            raise RuntimeError("No program compiled for this set of static arguments.") from e
+
+    def _param_index(self, name: str) -> int:
+        return list(self.program_type.definition.pos_or_kw_args.keys()).index(name)
+
+    @functools.cached_property
+    def _static_arg_indices(self) -> tuple[int, ...]:
+        if self.static_params is None:
+            # this could also be done in `__call__` but would be an extra if in the fast path
+            self.static_params = ()
+        return tuple(self._param_index(p) for p in self.static_params)
+
+    def _compile_variant(
+        self,
+        static_args: dict[str, ScalarOrTupleOfScalars],
+        offset_provider: common.OffsetProviderType | common.OffsetProvider,
+    ) -> None:
+        assert self.static_params is not None
+        static_args = _sanitize_static_args(
+            self.definition_stage.definition.__name__, static_args, self.program_type
+        )
+
+        args = tuple(
+            arguments.StaticArg(value=static_args[name], type_=type_)
+            if name in self.static_params
+            else type_
+            for name, type_ in self.program_type.definition.pos_or_kw_args.items()
+        )
+
+        compile_time_args = arguments.CompileTimeArgs(
+            offset_provider=offset_provider,  # type:ignore[arg-type] # TODO(havogt): resolve OffsetProviderType vs OffsetProvider
+            column_axis=None,  # TODO(havogt): column_axis seems to a unused, even for programs with scans
+            args=args,
+            kwargs={},
+        )
+        key = _CompiledProgramsKey(
+            tuple(static_args[p] for p in self.static_params),
+            offset_provider
+            if common.is_offset_provider_type(offset_provider)
+            else common.offset_provider_to_type(offset_provider),
+        )
+        if key in self._compiled_programs:
+            raise ValueError(f"Program with key {key} already exists.")
+        self._compiled_programs[key] = _async_compilation_pool.submit(
+            self.backend.compile, self.definition_stage, compile_time_args=compile_time_args
+        )
```

Comment:
Just an optional suggestion: compute `compile_time_args` just before is used:

```suggestion
        key = _CompiledProgramsKey(
            tuple(static_args[p] for p in self.static_params),
            offset_provider
            if common.is_offset_provider_type(offset_provider)
            else common.offset_provider_to_type(offset_provider),
        )
        if key in self._compiled_programs:
            raise ValueError(f"Program with key {key} already exists.")

        compile_time_args = arguments.CompileTimeArgs(
            offset_provider=offset_provider,  # type:ignore[arg-type] # TODO(havogt): resolve OffsetProviderType vs OffsetProvider
            column_axis=None,  # TODO(havogt): column_axis seems to a unused, even for programs with scans
            args=args,
            kwargs={},
        )            
        self._compiled_programs[key] = _async_compilation_pool.submit(
            self.backend.compile, self.definition_stage, compile_time_args=compile_time_args
        )
```